### PR TITLE
Remove legacy primary shard allocation mode based on versions

### PR DESCRIPTION
**Pull request by ywelsch**
Elasticsearch v5.0.0 uses allocation IDs to safely allocate primary shards whereas prior versions of ES used a version-based mode instead. Elasticsearch v5 still has support for version-based primary shard allocation as it needs to be able to load 2.x shards. ES v6 can drop the legacy support.

### CODE REVIEW

File: core/src/test/java/org/elasticsearch/gateway/PrimaryShardAllocatorTests.java [hunk 1]
```diff
@@ -119,17 +113,13 @@ public void testNoAsyncFetchData() {
...
      */
     public void testNoAllocationFound() {
-        final RoutingAllocation allocation;
-        if (randomBoolean()) {
-            allocation = routingAllocationWithOnePrimaryNoReplicas(yesAllocationDeciders(), CLUSTER_RECOVERED, Version.CURRENT, "allocId");
-        } else {
-            allocation = routingAllocationWithOnePrimaryNoReplicas(yesAllocationDeciders(), CLUSTER_RECOVERED, Version.V_2_1_0);
-        }
-        testAllocator.addData(node1, ShardStateMetaData.NO_VERSION, null, randomBoolean());
+        final RoutingAllocation allocation =
```
Comment (abeyad) [inline core/src/test/java/org/elasticsearch/gateway/PrimaryShardAllocatorTests.java]: The `routingAllocationWithOnePrimaryNoReplicas` method no longer needs to take a `Version` parameter, would be nice to get rid of it.

File: core/src/test/java/org/elasticsearch/indices/settings/UpdateSettingsIT.java [hunk 2]
```diff
@@ -191,20 +191,14 @@ public void testOpenCloseUpdateSettings() throws Exception {
...
-            // expected
-        }
+        client()
+            .admin()
+            .indices()
+            .prepareUpdateSettings("test")
+            .setSettings(Settings.builder().put(IndexMetaData.SETTING_NUMBER_OF_REPLICAS, 1))
+            .execute()
+            .actionGet();
+
```
Comment (abeyad) [inline core/src/test/java/org/elasticsearch/indices/settings/UpdateSettingsIT.java]: can we assert here that the number of replicas setting did indeed get updated?

**Review by abeyad — CHANGES_REQUESTED** [review]
I left a few comments, the stats on this PR are awesome :)

File: core/src/main/java/org/elasticsearch/gateway/TransportNodesListGatewayStartedShards.java [hunk 3]
```diff
@@ -257,31 +257,24 @@ public ShardId getShardId() {
...
-            this(node, legacyVersion, allocationId, primary, null);
+        public NodeGatewayStartedShards(DiscoveryNode node, String allocationId, boolean primary) {
+            this(node, allocationId, primary, null);
         }
 
-        public NodeGatewayStartedShards(DiscoveryNode node, long legacyVersion, String allocationId, boolean primary,
-                                        Exception storeException) {
+        public NodeGatewayStartedShards(DiscoveryNode node, String allocationId, boolean primary, Exception storeException) {
             super(node);
-            this.legacyVersion = legacyVersion;
```
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/gateway/TransportNodesListGatewayStartedShards.java]: No, it's also null if no shard info was found.

File: core/src/test/java/org/elasticsearch/indices/settings/UpdateSettingsIT.java (hunk 2 above)
Comment (ywelsch) [inline core/src/test/java/org/elasticsearch/indices/settings/UpdateSettingsIT.java]: sure, I've pushed 4617939

File: core/src/test/java/org/elasticsearch/gateway/PrimaryShardAllocatorTests.java (hunk 1 above)
Comment (ywelsch) [inline core/src/test/java/org/elasticsearch/gateway/PrimaryShardAllocatorTests.java]: sure, I've pushed 4617939

**Review by abeyad — APPROVED** [review]

**Comment by abeyad** [comment]
LGTM

**Comment by ywelsch** [comment]
Thanks @abeyad

**Pull request merged**